<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>the phydms program &mdash; phydms 2.0.4 documentation</title>
    
    <link rel="stylesheet" href="_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '2.0.4',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="phydms 2.0.4 documentation" href="index.html" />
    <link rel="next" title="auxiliary programs" href="auxiliary_progs.html" />
    <link rel="prev" title="how to use phydms" href="howtouse.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="auxiliary_progs.html" title="auxiliary programs"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="howtouse.html" title="how to use phydms"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">phydms 2.0.4 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="the-phydms-program">
<span id="phydms-prog"></span><h1><a class="toc-backref" href="#id1">the <code class="docutils literal"><span class="pre">phydms</span></code> program</a><a class="headerlink" href="#the-phydms-program" title="Permalink to this headline">¶</a></h1>
<div class="contents topic" id="contents">
<p class="topic-title first">Contents</p>
<ul class="simple">
<li><a class="reference internal" href="#the-phydms-program" id="id1">the <code class="docutils literal"><span class="pre">phydms</span></code> program</a><ul>
<li><a class="reference internal" href="#overview" id="id2">Overview</a></li>
<li><a class="reference internal" href="#command-line-usage" id="id3">Command-line usage</a></li>
<li><a class="reference internal" href="#output-files" id="id4">Output files</a><ul>
<li><a class="reference internal" href="#log-file" id="id5">Log file</a></li>
<li><a class="reference internal" href="#log-likelihood-file" id="id6">Log likelihood file</a></li>
<li><a class="reference internal" href="#model-parameters" id="id7">Model parameters</a></li>
<li><a class="reference internal" href="#tree-file" id="id8">Tree file</a></li>
<li><a class="reference internal" href="#site-specific-omega-file" id="id9">Site-specific omega file</a></li>
<li><a class="reference internal" href="#site-specific-differential-selection" id="id10">Site-specific differential selection</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="section" id="overview">
<h2><a class="toc-backref" href="#id2">Overview</a><a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal"><span class="pre">phydms</span></code> can perform phylogenetic analyses with <a class="reference internal" href="ExpCM.html#expcm"><span class="std std-ref">Experimentally Informed Codon Models</span></a> as well as with some standard non-site-specific substitution models (variants of the <em>YNGKP</em> models described in <a class="reference external" href="http://www.genetics.org/content/155/1/431">Yang, Nielsen, Goldman, and Krabbe Pederson, Genetics, 155:431-449</a>). In addition, <code class="docutils literal"><span class="pre">phydms</span></code> can be used to detect site-specific diversifying or differential selection using the <a class="reference internal" href="ExpCM.html#expcm"><span class="std std-ref">Experimentally Informed Codon Models</span></a> as described in <a class="reference external" href="http://biologydirect.biomedcentral.com/articles/10.1186/s13062-016-0172-z">Bloom, Biology Direct, 12:1</a>.</p>
<p>See below for information on the <a class="reference internal" href="#command-line-usage">Command-line usage</a> and <a class="reference internal" href="#output-files">Output files</a>.</p>
</div>
<div class="section" id="command-line-usage">
<h2><a class="toc-backref" href="#id3">Command-line usage</a><a class="headerlink" href="#command-line-usage" title="Permalink to this headline">¶</a></h2>
<p>Phylogenetic analysis informed by deep mutational scanning data. The &#8216;phydms&#8217; package is written by the Bloom lab (see <a class="reference external" href="https://github.com/jbloomlab/phydms/contributors">https://github.com/jbloomlab/phydms/contributors</a>). Version 2.0.4. Full documentation at <a class="reference external" href="http://jbloomlab.github.io/phydms">http://jbloomlab.github.io/phydms</a></p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">phydms</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">brlen</span> <span class="p">{</span><span class="n">scale</span><span class="p">,</span><span class="n">optimize</span><span class="p">}]</span> <span class="p">[</span><span class="o">--</span><span class="n">gammaomega</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">omegabysite</span><span class="p">]</span>
              <span class="p">[</span><span class="o">--</span><span class="n">omegabysite_fixsyn</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">diffprefsbysite</span><span class="p">]</span>
              <span class="p">[</span><span class="o">--</span><span class="n">diffprefsprior</span> <span class="n">DIFFPREFSPRIOR</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">fitphi</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">randprefs</span><span class="p">]</span>
              <span class="p">[</span><span class="o">--</span><span class="n">avgprefs</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">divpressure</span> <span class="n">DIVPRESSURE</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">ncpus</span> <span class="n">NCPUS</span><span class="p">]</span>
              <span class="p">[</span><span class="o">--</span><span class="n">fitprefsmethod</span> <span class="p">{</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">}]</span> <span class="p">[</span><span class="o">--</span><span class="n">ncats</span> <span class="n">NCATS</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">minbrlen</span> <span class="n">MINBRLEN</span><span class="p">]</span>
              <span class="p">[</span><span class="o">--</span><span class="n">minpref</span> <span class="n">MINPREF</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">seed</span> <span class="n">SEED</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">initparams</span> <span class="n">INITPARAMS</span><span class="p">]</span>
              <span class="p">[</span><span class="o">--</span><span class="n">profile</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">nograd</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">v</span><span class="p">]</span>
              <span class="n">alignment</span> <span class="n">tree</span> <span class="n">model</span> <span class="n">outprefix</span>
</pre></div>
</div>
<dl class="docutils">
<dt>Positional arguments:</dt>
<dd><table class="first last docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd>alignment</kbd></td>
<td><p class="first">Existing FASTA file with aligned codon sequences.</p>
<p>Should contain aligned DNA codon sequences.
Stop codons are <strong>not</strong> allowed, except if a stop codon is the terminal character in all sequences, in which case they are automatically trimmed.</p>
<p>No checking is done to remove identical sequences, so you may want to remove such sequences yourself if there a lot of them.</p>
<p class="last">The headers must be unique strings that do <strong>not</strong> contain any of the following: spaces, commas, colons, semicolons, parentheses, square brackets, and single or double quotes.</p>
</td></tr>
<tr><td class="option-group">
<kbd>tree</kbd></td>
<td><p class="first">Existing Newick file giving input tree.</p>
<p class="last">Typically you infer this tree using a program such as <a class="reference external" href="http://sco.h-its.org/exelixis/web/software/raxml/index.html">RAxML</a> or <a class="reference external" href="http://sourceforge.net/projects/codonphyml/">codonPhyML</a>.
The topology is fixed to that in <code class="docutils literal"><span class="pre">tree</span></code>. How the branch lengths are handled
during optimization depends on <code class="docutils literal"><span class="pre">--brlen</span></code>.
The branch lengths in <code class="docutils literal"><span class="pre">tree</span></code> are assumed to be in codon substitutions
per site.</p>
</td></tr>
<tr><td class="option-group">
<kbd>model</kbd></td>
<td><p class="first">Substitution model: ExpCM_&lt;prefsfile&gt; or YNGKP_&lt;m&gt; (where &lt;m&gt; is M0, M5). For ExpCM, &lt;prefsfile&gt; has first column labeled &#8216;site&#8217; and others labeled by 1-letter amino-acid code.</p>
<p><em>YNGKP_&lt;m&gt;</em> is a <em>non-site-specific</em> Goldman-Yang style codon model from <a class="reference external" href="http://www.genetics.org/content/155/1/431">Yang, Nielsen, Goldman, and Krabbe Pederson, Genetics, 155:431-449</a>.
The <em>&lt;m&gt;</em> indicates the model variant.
Codon frequencies are set empirically using the <em>CF3X4</em> method described by <a class="reference external" href="http://journals.plos.org/plosone/article?id=10.1371/journal.pone.0011230">Pond et al, PLoS One, 5:e11230</a> (this sets the codon frequencies to the the product of nucleotide frequency parameters at each of the three positions after correcting for stop codons).
The transition-transversion ratio <span class="math">\(\kappa\)</span> is optimized by maximum likelihood.
The nonsynonymous-synonymous ratio <span class="math">\(\omega\)</span> is optimized differently depending on the variant:</p>
<blockquote>
<div><ul class="simple">
<li><em>YNGKP_M0</em> : a single <span class="math">\(\omega\)</span> is optimized by maximum likelihood.</li>
<li><em>YNGKP_M5</em> : <span class="math">\(\omega\)</span> drawn from a gamma distribution with the two parameters optimized by maximum likelihood. The distribution has <code class="docutils literal"><span class="pre">--ncats</span></code> categories.</li>
</ul>
</div></blockquote>
<p><em>ExpCM_&lt;prefsfile&gt;</em> is an <a class="reference internal" href="ExpCM.html#expcm"><span class="std std-ref">Experimentally Informed Codon Models</span></a> with amino-acid preferences taken from the file <code class="docutils literal"><span class="pre">prefsfile</span></code>.
The preferences file must specify a preference for the amino acid encoded by every site in <code class="docutils literal"><span class="pre">alignment</span></code>, using sequential 1, 2, ... numbering.
Any stop-codon preferences specified in the file are ignored.
The preferences file can be in several formats:</p>
<blockquote>
<div><ul>
<li><p class="first">A comma-, tab-, or space-delimited file with the first column giving
the site and the rest giving the preference for each amino-acid:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">site</span> <span class="n">A</span> <span class="n">C</span> <span class="n">D</span> <span class="n">E</span> <span class="n">F</span> <span class="n">G</span> <span class="n">H</span> <span class="n">I</span> <span class="n">K</span> <span class="n">L</span> <span class="n">M</span> <span class="n">N</span> <span class="n">P</span> <span class="n">Q</span> <span class="n">R</span> <span class="n">S</span> <span class="n">T</span> <span class="n">V</span> <span class="n">W</span> <span class="n">Y</span>
<span class="mi">1</span> <span class="mf">0.044</span>  <span class="mf">0.046</span> <span class="mf">0.048</span> <span class="mf">0.0461</span> <span class="mf">0.046</span> <span class="mf">0.0411</span> <span class="mf">0.052</span> <span class="mf">0.027</span> <span class="mf">0.046</span> <span class="mf">0.043</span> <span class="mf">0.172</span> <span class="mf">0.045</span> <span class="mf">0.045</span> <span class="mf">0.045</span> <span class="mf">0.044</span> <span class="mf">0.041</span> <span class="mf">0.037</span> <span class="mf">0.028</span> <span class="mf">0.046</span> <span class="mf">0.048</span>
<span class="mi">2</span> <span class="mf">0.658</span> <span class="mf">0.006</span> <span class="mf">0.005</span> <span class="mf">0.029</span> <span class="mf">0.005</span> <span class="mf">0.019</span> <span class="mf">0.007</span> <span class="mf">0.004</span> <span class="mf">0.007</span> <span class="mf">0.001</span> <span class="mf">0.036</span> <span class="mf">0.005</span> <span class="mf">0.009</span> <span class="mf">0.003</span> <span class="mf">0.005</span> <span class="mf">0.014</span> <span class="mf">0.013</span> <span class="mf">0.148</span> <span class="mf">0.009</span> <span class="mf">0.005</span>
</pre></div>
</div>
</li>
<li><p class="first">The more complex <a class="reference external" href="http://jbloomlab.github.io/dms_tools">dms_tools</a> <a class="reference external" href="http://jbloomlab.github.io/dms_tools/fileformats.html#preferences-file">preferences file format</a></p>
</li>
</ul>
</div></blockquote>
<p class="last">Importantly, the amino-acid preferences must be obtained <strong>independently</strong> from the sequence
alignment being analyzed. A deep mutational scanning experiment is an independent means
of obtaining the preferences, but estimating them from the amino-acid frequencies in the alignment of homologs is not a valid approach as you are then estimating the preferences from the same sequences that you are subsequently analyzing.</p>
</td></tr>
<tr><td class="option-group">
<kbd>outprefix</kbd></td>
<td><p class="first">Output file prefix.</p>
<p>If this prefix contains a directory name, that directory is created if it does not already exist.</p>
<p class="last">See <a class="reference internal" href="#output-files">Output files</a> for a description of the files created with this prefix. Any existing files with these names are deleted at the start of the program&#8217;s execution.</p>
</td></tr>
</tbody>
</table>
</dd>
<dt>Options:</dt>
<dd><table class="first last docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--brlen=optimize</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td><p class="first">How to handle branch lengths: scale by single parameter or optimize each one</p>
<p>The tree topology is fixed to that in <code class="docutils literal"><span class="pre">tree</span></code>.
But there are several options for how the branch lengths are handled:</p>
<blockquote>
<div><ul class="simple">
<li><em>scale</em>: fix the relative length of the branches to the values in <code class="docutils literal"><span class="pre">tree</span></code>, but scale them all by a single parameter <span class="math">\(\mu\)</span> that is optimized. This approach will be faster.</li>
<li><em>optimize</em>: optimize all branch lengths as free parameters. This approach will be more accurate, but slower.</li>
</ul>
</div></blockquote>
<p class="last">Possible choices: scale, optimize</p>
</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--gammaomega=False</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td><p class="first">Omega for ExpCM from gamma distribution rather than single value. To achieve same for YNGKP, use &#8216;model&#8217; of YNGKP_M5.</p>
<p>Use this option for a <cite>model</cite> of <em>ExpCM</em> if you want <span class="math">\(\omega\)</span> to be drawn from <code class="docutils literal"><span class="pre">--ncats</span></code> gamma-distributed categories, similar to a <em>YNGKP_M5</em> model.
This option adds one extra free parameter, as we are now optimizing two parameters for the <span class="math">\(\omega\)</span> distribution. It will also increase run-time by about 5-fold if you use the default for <code class="docutils literal"><span class="pre">--ncats</span></code>.</p>
<p class="last">You do <strong>not</strong> use this option with <em>YNGKP</em> models; for those, simply set <code class="docutils literal"><span class="pre">model</span></code> to <em>YNGKP_M5</em> to get gamma-distributed <span class="math">\(\omega\)</span>.</p>
</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--omegabysite=False</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td><p class="first">Fit omega (dN/dS) for each site.</p>
<p class="last">If using a YNGKP model, then the <span class="math">\(\omega_r\)</span> value is nearly analogous that obtained using the <em>FEL</em> model described by <a class="reference external" href="http://mbe.oxfordjournals.org/content/22/5/1208.full">Kosakovsky Pond and Frost, Mol Biol Evol, 22:1208-1222</a>. If using and <em>ExpCM</em>, then <span class="math">\(\omega_r\)</span> has the meaning described in <a class="reference internal" href="ExpCM.html#expcm"><span class="std std-ref">Experimentally Informed Codon Models</span></a>. Essentially, we fix all other model / tree parameters and then compare a model that fits a synonymous and nonsynonymous rate to each site to a null model that only fits a synonymous rate; there is evidence for <span class="math">\(\omega_r \ne 1\)</span> if fitting both nonsynonymous and synonymous rate gives sufficiently better likelihood than fitting synonymous rate alone. See also the <code class="docutils literal"><span class="pre">--omegabysite_fixsyn</span></code> option.</p>
</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--omegabysite_fixsyn=False</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td><p class="first">For &#8216;&#8211;omegabysite&#8217;, assign all sites same dS rather than fit for each site.</p>
<p class="last">This option is meaningful only if you are using <code class="docutils literal"><span class="pre">--omegabysite</span></code>. If you use this option, then we compare a model in which we fit a nonsynonymous rate to each site to a model in which we fit nothing. The synonymous rate is not fit, and so is assumed to be equal to the overall value fit for the tree. According to <a class="reference external" href="http://mbe.oxfordjournals.org/content/22/5/1208.full">Kosakovsky Pond and Frost, Mol Biol Evol, 22:1208-1222</a>, in some cases this can yield greater power if there is relatively limited data. However, it comes with the risk of giving spurious results if there is substantial variation in the synonymous substitution rate among sites.</p>
</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--diffprefsbysite=False</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td><p class="first">Fit differential preferences for each site.</p>
<p>This option can only be used with <em>ExpCM</em> models, <strong>not</strong> with <em>YNGKP</em> models.</p>
<p>The fitting of site-specific differential selection is designed to identify sites that prefer different amino acids in the natural alignment than expected from the deep mutational scanning data.
This approach is described in <a class="reference external" href="http://biologydirect.biomedcentral.com/articles/10.1186/s13062-016-0172-z">Bloom, Biology Direct, 12:1</a>.</p>
<p class="last">The prior that is used is specified by <code class="docutils literal"><span class="pre">--diffprefsprior</span></code>.</p>
</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--diffprefsprior=invquadratic,150,0.5</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td><p class="first">Regularizing prior for &#8216;&#8211;diffprefsbysite&#8217;: &#8216;invquadratic,C1,C2&#8217; is prior in Bloom, Biology Direct, 12:1.</p>
<p>How do we regularize the differential preferences when using <code class="docutils literal"><span class="pre">--diffprefsbysite</span></code>?</p>
<p class="last">To use the inverse-quadratic prior in <a class="reference external" href="http://biologydirect.biomedcentral.com/articles/10.1186/s13062-016-0172-z">Bloom, Biology Direct, 12:1</a>, specify
the string <cite>invquadratic,C1,C2</cite> where <cite>C1</cite> and <cite>C2</cite> are numbers &gt; 0 that specify
how concentrated the prior is.</p>
</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--fitphi=False</span></kbd></td>
<td><p class="first">Fit ExpCM phi rather than setting so stationary state matches alignment frequencies.</p>
<p class="last">This option is not typically recommended. It will typically lead to only very slight improvements in log likelihood at substantial computational cost.</p>
</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--randprefs=False</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td><p class="first">Randomize preferences among sites for ExpCM.</p>
<p class="last">Only for <em>ExpCM</em> models.
This option randomly reassigns the preferences among sites.
This can be used as a control: randomizing the preferences should make them lose their efficacy for describing evolution.</p>
</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--avgprefs=False</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td><p class="first">Average preferences across sites for ExpCM.</p>
<p class="last">Only for <em>ExpCM</em> models.
This option computes an average of each preference across sites (<span class="math">\(\pi_a = \frac{1}{L} \sum_r \pi_{r,a}\)</span> where <span class="math">\(r = 1, \ldots, L\)</span>), and then uses these average preferences for all sites.
This can be used as a control, as it merges all the information in the preferences into a non-site-specific model.</p>
</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--divpressure</span></kbd></td>
<td><p class="first">Known diversifying pressure at sites: file with column 1 = position, column 2 = diversification pressure; columns space-, tab-, or comma-delimited.</p>
<p>Only for <em>ExpCM</em> models. This option specifies the name of the file with the predetermined diversifying pressure for each site, <span class="math">\(\delta_{r}\)</span>. <code class="docutils literal"><span class="pre">phydms</span></code> will fit <span class="math">\(\omega\)</span> and <span class="math">\(\omega_{2}\)</span> as described in <a class="reference internal" href="ExpCM.html#expcm"><span class="std std-ref">Experimentally Informed Codon Models</span></a>.</p>
<p class="last">You can <strong>not</strong> use <code class="docutils literal"><span class="pre">--divpressure</span></code> and <code class="docutils literal"><span class="pre">--fitphi</span></code> simultaneously.</p>
</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--ncpus=1</span></kbd></td>
<td><p class="first">Use this many CPUs; -1 means all available.</p>
<p class="last">Using multiples CPUs accelerates the by-site fitting (e.g., <code class="docutils literal"><span class="pre">--omegabysite</span></code>) as multiple sites can be analyzed at once.
However, it does not accelerate the initial optimization of the model and tree, as this calculation is not currently parallelized.</p>
</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--fitprefsmethod=2</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td><p class="first">Implementation to we use when fitting prefs.</p>
<p>As described in <a class="reference internal" href="implementation.html#implementation"><span class="std std-ref">Numerical implementation</span></a>, there are two different internal parameter representations that can be used when optimizing the preferences as free parameters.
Both in principle should give the same result, but in practice one method might be more efficient than the other or work better in rare cases where the optimization runs into problems.</p>
<p class="last">Possible choices: 1, 2</p>
</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--ncats=4</span></kbd></td>
<td><p class="first">Number of categories for gamma-distributed omega.</p>
<p class="last">Determines the number of categories when using <code class="docutils literal"><span class="pre">--gammaomega</span></code> and <cite>model</cite> of <em>YNGKP_M5</em>. More categories leads to longer run-time, values of 4-5 are usually adequate.</p>
</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--minbrlen=1e-06</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td><p class="first">Adjust all branch lengths in starting &#8216;tree&#8217; to &gt;= this.</p>
<p class="last">All branches with lengths less than this value will be set to this value in the initial starting tree.
Branches can still end up with lengths less than this after subsequent optimization of this starting tree.</p>
</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--minpref=0.002</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td><p class="first">Adjust all preferences in ExpCM &#8216;prefsfile&#8217; to &gt;= this.</p>
<p class="last">Adjust all amino-acid preferences in the <em>ExpCM</em> <code class="docutils literal"><span class="pre">prefsfile</span></code> to be at least this large.
If this number is too close to zero, you will get runtime errors.</p>
</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--seed=1</span></kbd></td>
<td><p class="first">Random number seed.</p>
<p class="last">The random number seed affects the outcomes when there is randomization, as when using <code class="docutils literal"><span class="pre">--randprefs</span></code>.</p>
</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--initparams</span></kbd></td>
<td>Initialize model params from this file, which should be format of &#8216;*_modelparams.txt&#8217; file created by &#8216;phydms&#8217; with this model.</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--profile=False</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Profile likelihood maximization, write pstats files. For code-development purposes.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--nograd=False</span></kbd></td>
<td>Do not use gradients for likelihood maximization.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">-v</span>, <span class="option">--version</span></kbd></td>
<td>show program&#8217;s version number and exit</td></tr>
</tbody>
</table>
</dd>
</dl>
</div>
<div class="section" id="output-files">
<h2><a class="toc-backref" href="#id4">Output files</a><a class="headerlink" href="#output-files" title="Permalink to this headline">¶</a></h2>
<p>Running <code class="docutils literal"><span class="pre">phydms</span></code> will create the following output files, all with the prefix specified by <code class="docutils literal"><span class="pre">outprefix</span></code>.</p>
<div class="section" id="log-file">
<h3><a class="toc-backref" href="#id5">Log file</a><a class="headerlink" href="#log-file" title="Permalink to this headline">¶</a></h3>
<p>This is a text file with the suffix <code class="docutils literal"><span class="pre">_log.log</span></code> that records information about the program&#8217;s progress.</p>
</div>
<div class="section" id="log-likelihood-file">
<h3><a class="toc-backref" href="#id6">Log likelihood file</a><a class="headerlink" href="#log-likelihood-file" title="Permalink to this headline">¶</a></h3>
<p>This file has the suffix <code class="docutils literal"><span class="pre">_loglikelihood.txt</span></code>. It simply gives the optimized log likelihood. Here is an example of a file&#8217;s contents:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">log</span> <span class="n">likelihood</span> <span class="o">=</span> <span class="o">-</span><span class="mf">4415.24</span>
</pre></div>
</div>
</div>
<div class="section" id="model-parameters">
<h3><a class="toc-backref" href="#id7">Model parameters</a><a class="headerlink" href="#model-parameters" title="Permalink to this headline">¶</a></h3>
<p>This file has the suffix <code class="docutils literal"><span class="pre">_modelparams.txt</span></code>.</p>
<p>Here is an example of the contents for an <em>ExpCM</em> model.
It gives the values of the parameters described in <a class="reference internal" href="ExpCM.html#expcm"><span class="std std-ref">Experimentally Informed Codon Models</span></a>; note that <span class="math">\(\mu\)</span> is <strong>not</strong> included as it is confounded with the branch lengths:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">beta</span> <span class="o">=</span> <span class="mf">2.99307</span>
<span class="n">kappa</span> <span class="o">=</span> <span class="mf">6.31364</span>
<span class="n">omega</span> <span class="o">=</span> <span class="mf">0.780782</span>
<span class="n">phiA</span> <span class="o">=</span> <span class="mf">0.372278</span>
<span class="n">phiC</span> <span class="o">=</span> <span class="mf">0.196416</span>
<span class="n">phiG</span> <span class="o">=</span> <span class="mf">0.224367</span>
</pre></div>
</div>
<p>Here is an example of the contents for a <em>YNGKP_M0</em> model.
In this file, <cite>phi0A</cite> is the corrected empirical frequency of <cite>A</cite> at the first codon site, <cite>phi0A</cite> is the frequency of <cite>A</cite> at the second codon site, etc.:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">kappa</span> <span class="o">=</span> <span class="mf">5.78484</span>
<span class="n">omega</span> <span class="o">=</span> <span class="mf">0.11308</span>
<span class="n">phi0A</span> <span class="o">=</span> <span class="mf">0.340656</span>
<span class="n">phi0C</span> <span class="o">=</span> <span class="mf">0.153395</span>
<span class="n">phi0G</span> <span class="o">=</span> <span class="mf">0.308294</span>
<span class="n">phi1A</span> <span class="o">=</span> <span class="mf">0.320503</span>
<span class="n">phi1C</span> <span class="o">=</span> <span class="mf">0.206698</span>
<span class="n">phi1G</span> <span class="o">=</span> <span class="mf">0.236844</span>
<span class="n">phi2A</span> <span class="o">=</span> <span class="mf">0.331485</span>
<span class="n">phi2C</span> <span class="o">=</span> <span class="mf">0.202991</span>
<span class="n">phi2G</span> <span class="o">=</span> <span class="mf">0.243997</span>
</pre></div>
</div>
<p>If you use a model with a gamma-distributed <span class="math">\(\omega\)</span> (i.e., the <code class="docutils literal"><span class="pre">--gammarates</span></code> option for an <em>ExpCM</em>, or the <em>YNGKP_M5</em> model), then rather than having a single <span class="math">\(\omega\)</span> value, there are instead two parameters that determine the gamma distribution.
These are the shape parameter <span class="math">\(\alpha_{\omega}\)</span> (denoted <em>alpha_omega</em>) and the inverse scale parameter <span class="math">\(\beta_{\omega}\)</span> (denoted by <em>beta_omega</em>).
The mean and variance of the omega distribution are <span class="math">\(\alpha_{\omega}/ \beta_{\omega}\)</span> and <span class="math">\(\alpha_{\omega} / \left(\beta_{\omega}\right)^2\)</span>, respectively.
To get the exact values, use the <a class="reference internal" href="api.html#api"><span class="std std-ref">phydmslib API</span></a> to call <code class="docutils literal"><span class="pre">phydmslib.models.DiscreteGamma(alpha_omega,</span> <span class="pre">beta_omega,</span> <span class="pre">ncats)</span></code> where <em>ncats</em> is the value set by <code class="docutils literal"><span class="pre">--ncats</span></code>. Here is an example of the model parameter file contents for an <em>ExpCM</em> with <code class="docutils literal"><span class="pre">--gammaomega</span></code>:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">alpha_omega</span> <span class="o">=</span> <span class="mf">0.835183</span>
<span class="n">beta</span> <span class="o">=</span> <span class="mf">3.01549</span>
<span class="n">beta_omega</span> <span class="o">=</span> <span class="mf">0.976053</span>
<span class="n">kappa</span> <span class="o">=</span> <span class="mf">6.3424</span>
<span class="n">phiA</span> <span class="o">=</span> <span class="mf">0.372475</span>
<span class="n">phiC</span> <span class="o">=</span> <span class="mf">0.196471</span>
<span class="n">phiG</span> <span class="o">=</span> <span class="mf">0.22418</span>
</pre></div>
</div>
</div>
<div class="section" id="tree-file">
<h3><a class="toc-backref" href="#id8">Tree file</a><a class="headerlink" href="#tree-file" title="Permalink to this headline">¶</a></h3>
<p>This file has the suffix <code class="docutils literal"><span class="pre">_tree.newick</span></code>, and gives the optimized tree in Newick format.
The branch lengths are in units of codon substitutions per site.</p>
</div>
<div class="section" id="site-specific-omega-file">
<h3><a class="toc-backref" href="#id9">Site-specific omega file</a><a class="headerlink" href="#site-specific-omega-file" title="Permalink to this headline">¶</a></h3>
<p>This file has the suffix <code class="docutils literal"><span class="pre">_omegabysite.txt</span></code>, and is created only if using the <code class="docutils literal"><span class="pre">--omegabysite</span></code> option. This file gives the <span class="math">\(\omega_r\)</span> values optimized for each site. In the case of a <em>YNGKP</em> model, these are site-specific dN/dS ratios that should be essentially analogous to those obtained under the <em>FEL</em> model described by <a class="reference external" href="http://mbe.oxfordjournals.org/content/22/5/1208.full">Kosakovsky Pond and Frost, Mol Biol Evol, 22:1208-1222</a>. In the case of an <em>ExpCM</em> model, these values indicate diversifying selection for nonsynonymous amino-acid change as described in <a class="reference external" href="http://biologydirect.biomedcentral.com/articles/10.1186/s13062-016-0172-z">Bloom, Biology Direct, 12:1</a> (see also <a class="reference internal" href="ExpCM.html#expcm"><span class="std std-ref">Experimentally Informed Codon Models</span></a>).</p>
<p>Here is an example of the first few lines of a file. The entries are tab separated:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">site</span>    <span class="n">omega</span>   <span class="n">P</span>   <span class="n">dLnL</span>    <span class="n">Q</span>
<span class="mi">213</span> <span class="mf">0.000</span>   <span class="mf">0.000216</span>    <span class="mf">6.841</span>   <span class="mf">0.108</span>
<span class="mi">55</span>  <span class="mf">0.000</span>   <span class="mf">0.000649</span>    <span class="mf">5.815</span>   <span class="mf">0.162</span>
<span class="mi">354</span> <span class="mf">0.000</span>   <span class="mf">0.00166</span> <span class="mf">4.948</span>   <span class="mf">0.177</span>
<span class="mi">298</span> <span class="mf">0.000</span>   <span class="mf">0.00181</span> <span class="mf">4.865</span>   <span class="mf">0.177</span>
</pre></div>
</div>
<p>The <em>P</em>-values are for rejection of the null hypothesis that <span class="math">\(\omega_r = 1\)</span> (calculated using a <span class="math">\(\chi_1^2\)</span> test).
These <em>P</em>-values are <strong>not</strong> corrected for multiple hypothesis testing; so you should instead look at the <em>Q</em>-values, which give the false discovery rate if this site and all above it are considered to have <span class="math">\(\omega \ne 1\)</span>.
The sites are sorted in the file by <em>P</em>-value.</p>
<p>For this type of test, the quantity of interest is typically not so much the exact value of <span class="math">\(\omega_r\)</span>, but rather the <em>P</em>-value (or <em>Q</em>-value) that <span class="math">\(\omega_r \ne 1\)</span>.
That <em>P</em>-value (or <em>Q</em>-value) is what indicates the credence that you should lend to the idea that a site is under diversifying selection if <span class="math">\(\omega_r &gt; 1\)</span>.</p>
</div>
<div class="section" id="site-specific-differential-selection">
<h3><a class="toc-backref" href="#id10">Site-specific differential selection</a><a class="headerlink" href="#site-specific-differential-selection" title="Permalink to this headline">¶</a></h3>
<p>This file has the suffix <code class="docutils literal"><span class="pre">_diffprefsbysite.txt</span></code>, and is created only if using the <code class="docutils literal"><span class="pre">--diffprefsbysite</span></code> option to an <em>ExpCM</em>. These are the differential preferences (the <span class="math">\(\Delta\pi_{r,a}\)</span> values) described in <a class="reference external" href="http://biologydirect.biomedcentral.com/articles/10.1186/s13062-016-0172-z">Bloom, Biology Direct, 12:1</a> and <a class="reference internal" href="ExpCM.html#expcm"><span class="std std-ref">Experimentally Informed Codon Models</span></a>.</p>
<p>Here is an example of the first few lines of a file. The entries are tab separated:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">site</span>    <span class="n">dpi_A</span>   <span class="n">dpi_C</span>   <span class="n">dpi_D</span>   <span class="n">dpi_E</span>   <span class="n">dpi_F</span>   <span class="n">dpi_G</span>   <span class="n">dpi_H</span>   <span class="n">dpi_I</span>   <span class="n">dpi_K</span>   <span class="n">dpi_L</span>   <span class="n">dpi_M</span>   <span class="n">dpi_N</span>   <span class="n">dpi_P</span>   <span class="n">dpi_Q</span>   <span class="n">dpi_R</span>   <span class="n">dpi_S</span>   <span class="n">dpi_T</span>   <span class="n">dpi_V</span>   <span class="n">dpi_W</span>   <span class="n">dpi_Y</span>   <span class="n">half_sum_abs_dpi</span>
<span class="mi">375</span> <span class="o">-</span><span class="mf">0.0004</span> <span class="o">-</span><span class="mf">0.0040</span> <span class="mf">0.0995</span>  <span class="o">-</span><span class="mf">0.8408</span> <span class="o">-</span><span class="mf">0.0014</span> <span class="mf">0.4839</span>  <span class="o">-</span><span class="mf">0.0029</span> <span class="o">-</span><span class="mf">0.0003</span> <span class="o">-</span><span class="mf">0.0062</span> <span class="o">-</span><span class="mf">0.0333</span> <span class="o">-</span><span class="mf">0.0092</span> <span class="o">-</span><span class="mf">0.0039</span> <span class="o">-</span><span class="mf">0.0003</span> <span class="o">-</span><span class="mf">0.0340</span> <span class="o">-</span><span class="mf">0.0000</span> <span class="o">-</span><span class="mf">0.0036</span> <span class="o">-</span><span class="mf">0.0073</span> <span class="mf">0.3659</span>  <span class="o">-</span><span class="mf">0.0013</span> <span class="o">-</span><span class="mf">0.0003</span> <span class="mf">0.9492</span>
<span class="mi">421</span> <span class="o">-</span><span class="mf">0.0107</span> <span class="o">-</span><span class="mf">0.0062</span> <span class="o">-</span><span class="mf">0.0092</span> <span class="mf">0.1521</span>  <span class="o">-</span><span class="mf">0.0009</span> <span class="o">-</span><span class="mf">0.0004</span> <span class="o">-</span><span class="mf">0.0078</span> <span class="o">-</span><span class="mf">0.0124</span> <span class="o">-</span><span class="mf">0.0003</span> <span class="o">-</span><span class="mf">0.0168</span> <span class="o">-</span><span class="mf">0.0036</span> <span class="o">-</span><span class="mf">0.0271</span> <span class="o">-</span><span class="mf">0.0000</span> <span class="o">-</span><span class="mf">0.0019</span> <span class="o">-</span><span class="mf">0.0060</span> <span class="o">-</span><span class="mf">0.0259</span> <span class="o">-</span><span class="mf">0.0022</span> <span class="o">-</span><span class="mf">0.0116</span> <span class="o">-</span><span class="mf">0.0003</span> <span class="o">-</span><span class="mf">0.0085</span> <span class="mf">0.1521</span>
<span class="mi">283</span> <span class="o">-</span><span class="mf">0.0126</span> <span class="o">-</span><span class="mf">0.0124</span> <span class="o">-</span><span class="mf">0.0001</span> <span class="o">-</span><span class="mf">0.0031</span> <span class="o">-</span><span class="mf">0.0052</span> <span class="o">-</span><span class="mf">0.0129</span> <span class="o">-</span><span class="mf">0.0131</span> <span class="o">-</span><span class="mf">0.0047</span> <span class="o">-</span><span class="mf">0.0014</span> <span class="o">-</span><span class="mf">0.0085</span> <span class="o">-</span><span class="mf">0.0077</span> <span class="o">-</span><span class="mf">0.0012</span> <span class="mf">0.1502</span>  <span class="o">-</span><span class="mf">0.0123</span> <span class="o">-</span><span class="mf">0.0004</span> <span class="o">-</span><span class="mf">0.0237</span> <span class="o">-</span><span class="mf">0.0164</span> <span class="o">-</span><span class="mf">0.0089</span> <span class="o">-</span><span class="mf">0.0054</span> <span class="o">-</span><span class="mf">0.0000</span> <span class="mf">0.1502</span>
<span class="mi">294</span> <span class="o">-</span><span class="mf">0.0078</span> <span class="o">-</span><span class="mf">0.0087</span> <span class="o">-</span><span class="mf">0.0112</span> <span class="mf">0.1458</span>  <span class="o">-</span><span class="mf">0.0084</span> <span class="o">-</span><span class="mf">0.0079</span> <span class="o">-</span><span class="mf">0.0098</span> <span class="o">-</span><span class="mf">0.0001</span> <span class="o">-</span><span class="mf">0.0064</span> <span class="o">-</span><span class="mf">0.0028</span> <span class="o">-</span><span class="mf">0.0079</span> <span class="o">-</span><span class="mf">0.0134</span> <span class="o">-</span><span class="mf">0.0000</span> <span class="o">-</span><span class="mf">0.0126</span> <span class="o">-</span><span class="mf">0.0104</span> <span class="o">-</span><span class="mf">0.0156</span> <span class="o">-</span><span class="mf">0.0110</span> <span class="o">-</span><span class="mf">0.0115</span> <span class="o">-</span><span class="mf">0.0001</span> <span class="o">-</span><span class="mf">0.0002</span> <span class="mf">0.1458</span>
<span class="mi">127</span> <span class="o">-</span><span class="mf">0.0088</span> <span class="o">-</span><span class="mf">0.0006</span> <span class="o">-</span><span class="mf">0.0010</span> <span class="mf">0.1423</span>  <span class="o">-</span><span class="mf">0.0021</span> <span class="o">-</span><span class="mf">0.0179</span> <span class="o">-</span><span class="mf">0.0059</span> <span class="o">-</span><span class="mf">0.0096</span> <span class="o">-</span><span class="mf">0.0208</span> <span class="o">-</span><span class="mf">0.0100</span> <span class="o">-</span><span class="mf">0.0021</span> <span class="o">-</span><span class="mf">0.0095</span> <span class="o">-</span><span class="mf">0.0007</span> <span class="o">-</span><span class="mf">0.0066</span> <span class="o">-</span><span class="mf">0.0073</span> <span class="o">-</span><span class="mf">0.0114</span> <span class="o">-</span><span class="mf">0.0146</span> <span class="o">-</span><span class="mf">0.0075</span> <span class="o">-</span><span class="mf">0.0010</span> <span class="o">-</span><span class="mf">0.0049</span> <span class="mf">0.1423</span>
<span class="mi">289</span> <span class="o">-</span><span class="mf">0.0079</span> <span class="o">-</span><span class="mf">0.0127</span> <span class="o">-</span><span class="mf">0.0005</span> <span class="o">-</span><span class="mf">0.0002</span> <span class="o">-</span><span class="mf">0.0228</span> <span class="o">-</span><span class="mf">0.0005</span> <span class="o">-</span><span class="mf">0.0154</span> <span class="o">-</span><span class="mf">0.0156</span> <span class="o">-</span><span class="mf">0.0033</span> <span class="o">-</span><span class="mf">0.0167</span> <span class="o">-</span><span class="mf">0.0113</span> <span class="o">-</span><span class="mf">0.0034</span> <span class="o">-</span><span class="mf">0.0004</span> <span class="o">-</span><span class="mf">0.0004</span> <span class="o">-</span><span class="mf">0.0094</span> <span class="o">-</span><span class="mf">0.0020</span> <span class="o">-</span><span class="mf">0.0028</span> <span class="o">-</span><span class="mf">0.0133</span> <span class="o">-</span><span class="mf">0.0006</span> <span class="mf">0.1391</span>  <span class="mf">0.1391</span>
</pre></div>
</div>
<p>The first column gives the site numbers, subsequent columns give the differential preference (<span class="math">\(\Delta\pi_{r,a}\)</span>) for each amino acid.
The last column gives the half absolute sum of the differential preferences, <span class="math">\(\sum_a |\Delta\pi_{r,a}|\)</span>, at each site. This quantity can range from zero to one.
The sites are sorted with the highest half absolute sum differential preference first.</p>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">the <code class="docutils literal"><span class="pre">phydms</span></code> program</a><ul>
<li><a class="reference internal" href="#overview">Overview</a></li>
<li><a class="reference internal" href="#command-line-usage">Command-line usage</a></li>
<li><a class="reference internal" href="#output-files">Output files</a><ul>
<li><a class="reference internal" href="#log-file">Log file</a></li>
<li><a class="reference internal" href="#log-likelihood-file">Log likelihood file</a></li>
<li><a class="reference internal" href="#model-parameters">Model parameters</a></li>
<li><a class="reference internal" href="#tree-file">Tree file</a></li>
<li><a class="reference internal" href="#site-specific-omega-file">Site-specific omega file</a></li>
<li><a class="reference internal" href="#site-specific-differential-selection">Site-specific differential selection</a></li>
</ul>
</li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="howtouse.html"
                        title="previous chapter">how to use <code class="docutils literal"><span class="pre">phydms</span></code></a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="auxiliary_progs.html"
                        title="next chapter">auxiliary programs</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/phydms_prog.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="auxiliary_progs.html" title="auxiliary programs"
             >next</a> |</li>
        <li class="right" >
          <a href="howtouse.html" title="how to use phydms"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">phydms 2.0.4 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2015, Jesse D. Bloom.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.4.1.
    </div>
  </body>
</html>