<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>the phydms_comprehensive program &mdash; phydms 2.0.dev0 documentation</title>
    
    <link rel="stylesheet" href="_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '2.0.dev0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="phydms 2.0.dev0 documentation" href="index.html" />
    <link rel="up" title="auxiliary programs" href="auxiliary_progs.html" />
    <link rel="next" title="the phydms_prepalignment program" href="phydms_prepalignment.html" />
    <link rel="prev" title="auxiliary programs" href="auxiliary_progs.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="phydms_prepalignment.html" title="the phydms_prepalignment program"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="auxiliary_progs.html" title="auxiliary programs"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">phydms 2.0.dev0 documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="auxiliary_progs.html" accesskey="U">auxiliary programs</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="the-phydms-comprehensive-program">
<span id="phydms-comprehensive-prog"></span><h1><a class="toc-backref" href="#id1">the <code class="docutils literal"><span class="pre">phydms_comprehensive</span></code> program</a><a class="headerlink" href="#the-phydms-comprehensive-program" title="Permalink to this headline">¶</a></h1>
<div class="contents topic" id="contents">
<p class="topic-title first">Contents</p>
<ul class="simple">
<li><a class="reference internal" href="#the-phydms-comprehensive-program" id="id1">the <code class="docutils literal"><span class="pre">phydms_comprehensive</span></code> program</a><ul>
<li><a class="reference internal" href="#overview" id="id2">Overview</a></li>
<li><a class="reference internal" href="#command-line-usage" id="id3">Command-line usage</a></li>
<li><a class="reference internal" href="#output-files" id="id4">Output files</a><ul>
<li><a class="reference internal" href="#log-file" id="id5">Log file</a></li>
<li><a class="reference internal" href="#model-comparison-files" id="id6">Model comparison files</a></li>
<li><a class="reference internal" href="#phydms-output-for-each-model" id="id7"><code class="docutils literal"><span class="pre">phydms</span></code> output for each model</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="section" id="overview">
<h2><a class="toc-backref" href="#id2">Overview</a><a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal"><span class="pre">phydms_comprehensive</span></code> is a program that simplifies usage of <code class="docutils literal"><span class="pre">phydms</span></code> for standard analyses. Essentially, <code class="docutils literal"><span class="pre">phydms_comprehensive</span></code> runs <code class="docutils literal"><span class="pre">phydms</span></code> for several different models to enable model comparisons and identify selection.</p>
<p>In its simplest usage, you simply provide <code class="docutils literal"><span class="pre">phydms_comprehensive</span></code> with an alignment and one or more files giving site-specific amino-acid preferences.
The program then first uses <code class="docutils literal"><span class="pre">RAxML</span></code> to infer a tree under the <em>GTRCAT</em> model.
Alternatively, you can specify the tree using <code class="docutils literal"><span class="pre">--tree</span></code> flag.
For each set of site-specific amino-acid preferences, <code class="docutils literal"><span class="pre">phydms_comprehensive</span></code> optimizes the tree with the following models:</p>
<blockquote>
<div><ul class="simple">
<li><em>ExpCM</em></li>
<li><em>ExpCM</em> with a gamma-distributed <span class="math">\(\omega\)</span> (if using the <code class="docutils literal"><span class="pre">--gammaomega</span></code> flag).</li>
<li><em>YNGKP_M0</em></li>
<li><em>YNGKP_M5</em></li>
</ul>
</div></blockquote>
<p>It also runs each of these <em>ExpCM</em> with averaged preferences as a control.
Finally, it creates summaries that enable comparison among the models.</p>
<p>You could get all of this output by simply running <code class="docutils literal"><span class="pre">phydms</span></code> repeatedly, but using <code class="docutils literal"><span class="pre">phydms_comprehensive</span></code> automates this process for you.</p>
<p>See below for information on <a class="reference internal" href="#command-line-usage">Command-line usage</a> and <a class="reference internal" href="#output-files">Output files</a>.</p>
</div>
<div class="section" id="command-line-usage">
<h2><a class="toc-backref" href="#id3">Command-line usage</a><a class="headerlink" href="#command-line-usage" title="Permalink to this headline">¶</a></h2>
<p>Comprehensive phylogenetic model comparison and detection of selection informed by deep mutational scanning data. This program runs &#8216;phydms&#8217; repeatedly to compare substitution models and detect selection. The &#8216;phydms&#8217; package is written by the Bloom lab (see <a class="reference external" href="https://github.com/jbloomlab/phydms/contributors">https://github.com/jbloomlab/phydms/contributors</a>). Version 2.0.dev0. Full documentation at <a class="reference external" href="http://jbloomlab.github.io/phydms">http://jbloomlab.github.io/phydms</a></p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">phydms_comprehensive</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">(</span><span class="o">--</span><span class="n">raxml</span> <span class="n">RAXML</span> <span class="o">|</span> <span class="o">--</span><span class="n">tree</span> <span class="n">TREE</span><span class="p">)</span> <span class="p">[</span><span class="o">--</span><span class="n">ncpus</span> <span class="n">NCPUS</span><span class="p">]</span>
                            <span class="p">[</span><span class="o">--</span><span class="n">brlen</span> <span class="p">{</span><span class="n">scale</span><span class="p">,</span><span class="n">optimize</span><span class="p">}]</span> <span class="p">[</span><span class="o">--</span><span class="n">omegabysite</span><span class="p">]</span>
                            <span class="p">[</span><span class="o">--</span><span class="n">diffprefsbysite</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">gammaomega</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">no</span><span class="o">-</span><span class="n">avgprefs</span><span class="p">]</span>
                            <span class="p">[</span><span class="o">--</span><span class="n">randprefs</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">v</span><span class="p">]</span>
                            <span class="n">outprefix</span> <span class="n">alignment</span> <span class="n">prefsfiles</span> <span class="p">[</span><span class="n">prefsfiles</span> <span class="o">...</span><span class="p">]</span>
</pre></div>
</div>
<dl class="docutils">
<dt>Positional arguments:</dt>
<dd><table class="first last docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd>outprefix</kbd></td>
<td><p class="first">Output file prefix.</p>
<p class="last">This prefix can be a directory name (e.g. <code class="docutils literal"><span class="pre">my_directory/</span></code>) if you want to create a new directory. See <a class="reference internal" href="#output-files">Output files</a> for a description of the created files.</p>
</td></tr>
<tr><td class="option-group">
<kbd>alignment</kbd></td>
<td><p class="first">Existing FASTA file with aligned codon sequences.</p>
<p class="last">The alignment must meet the same specifications described in the <code class="docutils literal"><span class="pre">phydms</span></code> documentation for the argument of the same name (see <a class="reference internal" href="phydms_prog.html#phydms-prog"><span class="std std-ref">the phydms program</span></a>).</p>
</td></tr>
<tr><td class="option-group">
<kbd>prefsfiles</kbd></td>
<td><p class="first">Existing files with site-specific amino-acid preferences.</p>
<p class="last">Provide the name of one or more files giving site-specific amino-acid preferences. These files should meet the same specifications described in the <code class="docutils literal"><span class="pre">phydms</span></code> documentation for the <em>prefsfile</em> that should accompany an <em>ExpCM</em> (see <a class="reference internal" href="phydms_prog.html#phydms-prog"><span class="std std-ref">the phydms program</span></a>).</p>
</td></tr>
</tbody>
</table>
</dd>
<dt>Options:</dt>
<dd><table class="first last docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd><span class="option">--raxml</span></kbd></td>
<td><p class="first">Path to RAxML (e.g., &#8216;raxml&#8217;)</p>
<p class="last">By default, <code class="docutils literal"><span class="pre">phydms_comprehensive</span></code> uses <code class="docutils literal"><span class="pre">RAxML</span></code> to infer a tree topology under the <em>Jukes Cantor</em> model and the command <code class="docutils literal"><span class="pre">raxml</span></code>. Use this argument to specify a path to <code class="docutils literal"><span class="pre">RAxML</span></code> other than <code class="docutils literal"><span class="pre">raxml</span></code>.  The tree inferred by <code class="docutils literal"><span class="pre">RAxML</span></code> can be found in the same directory as the other output files.</p>
</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--tree</span></kbd></td>
<td><p class="first">Existing Newick file giving input tree.</p>
<p class="last">If you want to instead fix the tree to some existing topology, use this argument and provide the name of a file giving a valid tree in Newick format. You cannot specify both <code class="docutils literal"><span class="pre">--tree</span></code> and <code class="docutils literal"><span class="pre">--raxml</span></code>.</p>
</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--ncpus=-1</span></kbd></td>
<td>Use this many CPUs; -1 means all available.</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--brlen=optimize</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td><p class="first">How to handle branch lengths: scale by single parameter or optimize each one</p>
<p class="last">Possible choices: scale, optimize</p>
</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--omegabysite=False</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Fit omega (dN/dS) for each site.</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--diffprefsbysite=False</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Fit differential preferences for each site.</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--gammaomega=False</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Fit ExpCM with gamma distributed omega.</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--no-avgprefs=False</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>No fitting of models with preferences averaged across sites for ExpCM.</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--randprefs=False</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Include ExpCM models with randomized preferences.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">-v</span>, <span class="option">--version</span></kbd></td>
<td>show program&#8217;s version number and exit</td></tr>
</tbody>
</table>
</dd>
</dl>
</div>
<div class="section" id="output-files">
<h2><a class="toc-backref" href="#id4">Output files</a><a class="headerlink" href="#output-files" title="Permalink to this headline">¶</a></h2>
<p>Running <code class="docutils literal"><span class="pre">phydms_comprehensive</span></code> will create the following output files, all with the prefix specified by <code class="docutils literal"><span class="pre">outprefix</span></code>.</p>
<div class="section" id="log-file">
<h3><a class="toc-backref" href="#id5">Log file</a><a class="headerlink" href="#log-file" title="Permalink to this headline">¶</a></h3>
<p>A file with the suffix <code class="docutils literal"><span class="pre">.log</span></code> will be created that summarizes the overall progress of <code class="docutils literal"><span class="pre">phydms_comprehensive</span></code>. If <code class="docutils literal"><span class="pre">outprefix</span></code> is just a directory name, this file will be called <code class="docutils literal"><span class="pre">log.log</span></code>.</p>
</div>
<div class="section" id="model-comparison-files">
<h3><a class="toc-backref" href="#id6">Model comparison files</a><a class="headerlink" href="#model-comparison-files" title="Permalink to this headline">¶</a></h3>
<p>A file with the suffix <code class="docutils literal"><span class="pre">modelcomparison.md</span></code> will be created that summarizes the model comparison.
For each model, it reports the <span class="math">\(\Delta\rm{AIC}\)</span>, the optimized log likelihood, and the values of key parameters.
This file is in Markdown format:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="o">|</span> <span class="n">Model</span>                   <span class="o">|</span> <span class="n">deltaAIC</span> <span class="o">|</span> <span class="n">LogLikelihood</span> <span class="o">|</span> <span class="n">nParams</span> <span class="o">|</span> <span class="n">ParamValues</span>                                   <span class="o">|</span>
<span class="o">|-------------------------|----------|---------------|---------|-----------------------------------------------|</span>
<span class="o">|</span> <span class="n">ExpCM_NP_prefs</span>          <span class="o">|</span> <span class="mf">0.00</span>     <span class="o">|</span> <span class="o">-</span><span class="mf">3389.38</span>      <span class="o">|</span> <span class="mi">6</span>       <span class="o">|</span> <span class="n">beta</span><span class="o">=</span><span class="mf">2.99</span><span class="p">,</span> <span class="n">kappa</span><span class="o">=</span><span class="mf">6.31</span><span class="p">,</span> <span class="n">omega</span><span class="o">=</span><span class="mf">0.78</span>             <span class="o">|</span>
<span class="o">|</span> <span class="n">averaged_ExpCM_NP_prefs</span> <span class="o">|</span> <span class="mf">2586.44</span>  <span class="o">|</span> <span class="o">-</span><span class="mf">4682.60</span>      <span class="o">|</span> <span class="mi">6</span>       <span class="o">|</span> <span class="n">beta</span><span class="o">=</span><span class="mf">0.28</span><span class="p">,</span> <span class="n">kappa</span><span class="o">=</span><span class="mf">6.51</span><span class="p">,</span> <span class="n">omega</span><span class="o">=</span><span class="mf">0.12</span>             <span class="o">|</span>
<span class="o">|</span> <span class="n">YNGKP_M5</span>                <span class="o">|</span> <span class="mf">2599.70</span>  <span class="o">|</span> <span class="o">-</span><span class="mf">4683.23</span>      <span class="o">|</span> <span class="mi">12</span>      <span class="o">|</span> <span class="n">alpha_omega</span><span class="o">=</span><span class="mf">0.30</span><span class="p">,</span> <span class="n">beta_omega</span><span class="o">=</span><span class="mf">2.41</span><span class="p">,</span> <span class="n">kappa</span><span class="o">=</span><span class="mf">5.84</span> <span class="o">|</span>
<span class="o">|</span> <span class="n">YNGKP_M0</span>                <span class="o">|</span> <span class="mf">2679.50</span>  <span class="o">|</span> <span class="o">-</span><span class="mf">4724.13</span>      <span class="o">|</span> <span class="mi">11</span>      <span class="o">|</span> <span class="n">kappa</span><span class="o">=</span><span class="mf">5.79</span><span class="p">,</span> <span class="n">omega</span><span class="o">=</span><span class="mf">0.11</span>                        <span class="o">|</span>
</pre></div>
</div>
</div>
<div class="section" id="phydms-output-for-each-model">
<h3><a class="toc-backref" href="#id7"><code class="docutils literal"><span class="pre">phydms</span></code> output for each model</a><a class="headerlink" href="#phydms-output-for-each-model" title="Permalink to this headline">¶</a></h3>
<p>For each individual model, there will also be all of the expected <code class="docutils literal"><span class="pre">phydms</span></code> output files as described in <a class="reference internal" href="phydms_prog.html#phydms-prog"><span class="std std-ref">the phydms program</span></a>. These files will begin with the prefix specified by <code class="docutils literal"><span class="pre">outprefix</span></code>, which will be followed by the name of the model.</p>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">the <code class="docutils literal"><span class="pre">phydms_comprehensive</span></code> program</a><ul>
<li><a class="reference internal" href="#overview">Overview</a></li>
<li><a class="reference internal" href="#command-line-usage">Command-line usage</a></li>
<li><a class="reference internal" href="#output-files">Output files</a><ul>
<li><a class="reference internal" href="#log-file">Log file</a></li>
<li><a class="reference internal" href="#model-comparison-files">Model comparison files</a></li>
<li><a class="reference internal" href="#phydms-output-for-each-model"><code class="docutils literal"><span class="pre">phydms</span></code> output for each model</a></li>
</ul>
</li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="auxiliary_progs.html"
                        title="previous chapter">auxiliary programs</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="phydms_prepalignment.html"
                        title="next chapter">the <code class="docutils literal"><span class="pre">phydms_prepalignment</span></code> program</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/phydms_comprehensive_prog.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="phydms_prepalignment.html" title="the phydms_prepalignment program"
             >next</a> |</li>
        <li class="right" >
          <a href="auxiliary_progs.html" title="auxiliary programs"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">phydms 2.0.dev0 documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="auxiliary_progs.html" >auxiliary programs</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2015, Jesse D. Bloom.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.4.1.
    </div>
  </body>
</html>